c++/src/Int128.cc

```diff
@@ -435,6 +435,13 @@ namespace orc {
     return buf.str();
   }
 
+  double Int128::toDouble() const {
+    if (fitsInLong()) {
+      return static_cast<double>(toLong());
+    }
+    return static_cast<double>(lowbits) + std::ldexp(static_cast<double>(highbits), 64);
```

Comment:
Should we do some error handling? What if the double value does not precisely store the original value?

Reply:
> 
I didn't check the precision because I want Int128::toDouble() to act like 
static_csat<double>(__int128_t). Do you think is that OK?

Reply:
Does the java code do the same thing?

Reply:
> Does the java code do the same thing?

The java side uses BigDecimal.doubleValue() to convert a decimal to a double.This operation would lose information, but it seems that the java ORC didn't do any checks too. 
```java
    if(intCompact != -9223372036854775808) {
            if (scale == 0) {
                return (double)intCompact;
            } else {
                /*
                 * If both intCompact and the scale can be exactly
                 * represented as double values, perform a single
                 * double multiply or divide to compute the (properly
                 * rounded) result.
                 */
                if (Math.abs(intCompact) < 1L<<52 ) {
                    // Don't have too guard against
                    // Math.abs(MIN_VALUE) because of outer check
                    // against INFLATED.
                    if (scale > 0 && scale < double10pow.length) {
                        return (double)intCompact / double10pow[scale];
                    } else if (scale < 0 && scale > -double10pow.length) {
                        return (double)intCompact * double10pow[-scale];
                    }
                }
            }
        }
        // Somewhat inefficient, but guaranteed to work.
        return Double.parseDouble(this.toString());
```

Reply:
It is OK to lose precision as IEEE754 is unable to store all integers as is.